packages/apputils-extension/src/shortcuts.tsx

```diff
@@ -0,0 +1,122 @@
+  trans: TranslationBundle,
+  activeElement?: Element
+) {
+  const { commands } = app;
+  const elt = activeElement ?? document.activeElement;
+
+  /**
+   * Find the distance from the target node to the first matching node.
+   *
+   * Based on Lumino private function commands.Private.targetDistance
+   * This traverses the DOM path from `elt` to the root
+   * computes the distance from `elt` to the first node which matches
+   * the CSS selector. If no match is found, `-1` is returned.
+   *
+   * It also stops traversal if the `data-lm-suppress-shortcuts` or
+   * `data-p-suppress-shortcuts` attributes are found.
+   */
+  function matchDistance(selector: string, elt: Element | null): number {
+    let targ = elt;
+    for (
+      let dist = 0;
+      targ !== null && targ !== targ.parentElement;
+      targ = targ.parentElement, ++dist
+    ) {
+      if (targ.hasAttribute('data-lm-suppress-shortcuts')) {
+        return -1;
+      }
+      if (targ.matches(selector)) {
+        return dist;
+      }
+    }
+    return -1;
+  }
+
+  // Find active keybindings for target element
+  const activeBindings = new Map<
+    string,
+    [number, CommandRegistry.IKeyBinding]
+  >();
+  for (let i = 0; i < commands.keyBindings.length; i++) {
+    const kb = commands.keyBindings[i];
+    let distance = matchDistance(kb.selector, elt);
+    if (distance < 0) {
+      continue;
+    }
+    let formatted = CommandRegistry.formatKeystroke(kb.keys);
+    if (activeBindings.has(formatted)) {
+      let oldBinding = activeBindings.get(formatted)!;
+      // if the existing binding takes precedence, ignore this binding by continuing
+      if (
+        oldBinding[0] < distance ||
+        (oldBinding[0] === distance &&
+          Selector.calculateSpecificity(oldBinding[1].selector) >
+            Selector.calculateSpecificity(kb.selector))
+      ) {
+        continue;
+      }
+    }
+    activeBindings.set(formatted, [distance, kb]);
+  }
+
+  // Group shortcuts by distance
+  let maxDistance = -1;
+  const groupedBindings = new Map<number, CommandRegistry.IKeyBinding[]>();
+  for (let [distance, binding] of activeBindings.values()) {
+    maxDistance = Math.max(distance, maxDistance);
+    if (!groupedBindings.has(distance)) {
+      groupedBindings.set(distance, []);
+    }
+    groupedBindings.get(distance)!.push(binding);
+  }
+
+  // Display shortcuts by group
+  const bindingTable: any = [];
+  for (let d = 0; d <= maxDistance; d++) {
+    if (groupedBindings.has(d)) {
+      bindingTable.push(
+        groupedBindings.get(d)!.map(b => (
+          <tr className="jp-ContextualShortcut-TableRow" key={b.command}>
+            <td>{commands.label(b.command, b.args)}</td>
+            <td className="jp-ContextualShortcut-TableItem">
+              {CommandRegistry.formatKeystroke(b.keys)}
+            </td>
+          </tr>
+        ))
+      );
+      bindingTable.push(
+        <tr className="jp-ContextualShortcut-TableLastRow"></tr>
```

Comment:
@andrii-i Thanks for the suggestion. I don't know where exactly I should declare these caps global const (in my case, the css classes I defined were in a .tsx file).  And in the example I saw, let's say we have a `const NAME_CLASS = 'jp-Name'`, then in the constructor of the related class, a `this.addClass(NAME_CLASS). Sorry to ask for this, I don't have experience with that.

Reply:
@HaudinFlorence No worries, sorry that I was not clear. By "[all caps](https://en.wikipedia.org/wiki/All_caps)" I meant exactly what you understood, that all letters of the constant name are capital letters. The idea is that instead of using arbitrary strings in different places, you declare all arbitrary constants (css class names, resolutions, endpoints IPs and so on) in one place (below import statements, outside of any function), add descriptive comments to them, and then use these constants in your code. This makes code easier to understand, makes it possible to reuse constants if needed).

You can follow example I linked above and declare it in in the same file you use it in, outside of scope of any function, right below imports. For example you can do it at line 12 of packages/apputils-extension/src/shortcuts.tsx. It's also a good idea to add a comment describing the purpose of this class Something like:
```ts
/**
 * The class name added to last row of each ContextShortcutTable to ***insert purpose here***.
 */
const CONTEXT_SHORTCUT_TABLE_LAST_ROW_CLASS = 'jp-ContextualShortcut-TableLastRow'
```

Then you just use this constant wherever you would use the string before. Line 102 would then be:
```ts
        <tr className="jp-ContextualShortcut-TableLastRow"></tr>
```

Reply:
@andrii-i Thanks a lot for the precisions. Changes have been included to follow your comments.

Reply:
@HaudinFlorence Thank you for taking this and other suggestions into consideration